File: ocaml/hamming/hamming.ml
type nucleotide = A | C | G | T

let distance acc x y = acc + if x != y then 1 else 0

let hamming_distance xs ys =
  match (List.length xs, List.length ys) with
  | 0, 0 -> Ok 0
  | 0, _ -> Error "left strand must not be empty"
  | _, 0 -> Error "right strand must not be empty"
  | xlen, ylen ->
      if xlen != ylen then
        Error "left and right strands must be of equal length"
      else Ok (List.fold_left2 distance 0 xs ys)

File: erlang/hamming/src/hamming.erl
-module(hamming).

-export([distance/2]).

dist_of(Same, Same) -> 0;
dist_of(_, _) -> 1.

distance(Left, Right) ->
  case length(Left) =:= length(Right) of
    true -> lists:sum(lists:zipwith(fun dist_of/2, Left, Right));
    false -> {error, "left and right strands must be of equal length"}
  end.

